src/cp/generalizedhucocks.cxx

```diff
@@ -0,0 +1,836 @@
+    double d_pad_d_N = pmodel_->d_precipitate_areal_density_d_N(history, p);
+
+    // Third block: r
+    res.get<double>(p->varnames()[0]) = (0.5 * tau_c / svd * d_svd_d_f * p->d_f_d_r(history) +
+                                         0.5 * tau_p * tau_p / tau_s / pad * d_pad_d_r) *
+                                        p->rs();
+
+    // Fourth block: N
+    res.get<double>(p->varnames()[1]) = (0.5 * tau_c / svd * d_svd_d_f * p->d_f_d_N(history) +
+                                         0.5 * tau_p * tau_p / tau_s / pad * d_pad_d_N) *
+                                        p->Ns();
+  }
+
+  return res;
+}
+
+History
+GeneralizedHuCocksHardening::hist(const Symmetric & stress,
+                                  const Orientation & Q,
+                                  const History & history,
+                                  Lattice & L,
+                                  double T,
+                                  const SlipRule & R,
+                                  const History & fixed) const
+{
+  // Vector of results
+  History res = blank_hist().zero();
+
+  // History is easy, we just concatenate the dislocation model hardening
+  // followed by each precipitation hardening
+  auto h1 = dmodel_->hist(stress, Q, history, L, T, R, fixed);
+  std::copy(h1.rawptr(), h1.rawptr() + h1.size(), res.start_loc(dmodel_->varnames()[0]));
+
+  auto rate = pmodel_->rate(T, history);
+  for (unsigned int i = 0; i < pmodel_->varnames().size(); i++)
+    res.get<double>(pmodel_->varnames()[i]) = rate[i];
+
+  return res;
+}
+
+History
+GeneralizedHuCocksHardening::d_hist_d_s(const Symmetric & stress,
+                                        const Orientation & Q,
+                                        const History & history,
+                                        Lattice & L,
+                                        double T,
+                                        const SlipRule & R,
+                                        const History & fixed) const
+{
+  // This could be non-zero
+  History res = dmodel_->d_hist_d_s(stress, Q, history, L, T, R, fixed);
+
+  // These are zero
+  History pm;
+  pmodel_->populate_hist(pm);
+  res.add_union(pm.derivative<Symmetric>().zero());
+
+  return res;
+}
+
+History
+GeneralizedHuCocksHardening::d_hist_d_h(const Symmetric & stress,
+                                        const Orientation & Q,
+                                        const History & history,
+                                        Lattice & L,
+                                        double T,
+                                        const SlipRule & R,
+                                        const History & fixed) const
+{
+  // Start with the self-derivative
+  auto res = dmodel_->d_hist_d_h(stress, Q, history, L, T, R, fixed);
+
+  // pmodel/dmodel cross term
+  History phist;
+  pmodel_->populate_hist(phist);
+  res.add_union(phist.history_derivative(dmodel_->blank_hist()).zero());
+
+  // dmodel/pmodel cross term
```

Comment:
Why is this commented out?

Reply:
I believe this comment also exists in the original Hu-Cocks model. I left the comment here to remind myself to ask you why it was commented out.

Reply:
Great.  I have no ideas...